psi4/driver/procrouting/empirical_dispersion.py

```diff
@@ -25,259 +25,243 @@
+        molecule : psi4.core.Molecule
+            System for which to compute empirical dispersion correction.
+
+        Returns
+        -------
+        float
+            Dispersion energy [Eh].
+
+        """
+        if self.engine == 'dftd3':
+            jobrec = intf_dftd3.run_dftd3_from_arrays(
+                molrec=molecule.to_dict(np_out=False),
+                name_hint=self.fctldash,
+                level_hint=self.dashlevel,
+                param_tweaks=self.dashparams,
+                dashcoeff_supplement=self.dashcoeff_supplement,
+                ptype='energy',
+                verbose=1)
+
+            dashd_part = float(jobrec['qcvars']['DISPERSION CORRECTION ENERGY'].data)
+            for k, qca in jobrec['qcvars'].items():
+                if not isinstance(qca.data, np.ndarray):
+                    core.set_variable(k, qca.data)
+
+            if self.fctldash in ['hf3c', 'pbeh3c']:
+                gcp_part = gcp.run_gcp(molecule, self.fctldash, verbose=False, dertype=0)
+                dashd_part += gcp_part
+
+            return dashd_part
         else:
             return self.disp.compute_energy(molecule)
+            # TODO need to set 'DISPERSION CORRECTION ENERGY' and self.fctldash + 'DISPERSION CORRECTION ENERGY' psivars for libdisp?
 
     def compute_gradient(self, molecule):
-        if self.disp_type == 'gr':
-            if self.alias in ['hf3c', 'pbeh3c']:
-                dashd_part = dftd3.run_dftd3(
-                    molecule,
-                    dashlvl=self.dtype.lower().replace('-', ''),
-                    dashparam=self.dash_params,
-                    verbose=False,
-                    dertype=1)
-                gcp_part = gcp.run_gcp(molecule, self.alias.lower(), verbose=False, dertype=1)
+        """Compute dispersion gradient based on engine, dispersion level, and parameters in `self`.
+
+        Parameters
+        ----------
+        molecule : psi4.core.Molecule
+            System for which to compute empirical dispersion correction.
+
+        Returns
+        -------
+        psi4.core.Matrix
+            (nat, 3) dispersion gradient [Eh/a0].
+
+        """
+        if self.engine == 'dftd3':
+            jobrec = intf_dftd3.run_dftd3_from_arrays(
+                molrec=molecule.to_dict(np_out=False),
+                name_hint=self.fctldash,
+                level_hint=self.dashlevel,
+                param_tweaks=self.dashparams,
+                dashcoeff_supplement=self.dashcoeff_supplement,
+                ptype='gradient',
+                verbose=1)
+
+            dashd_part = core.Matrix.from_array(jobrec['qcvars']['DISPERSION CORRECTION GRADIENT'].data)
+            for k, qca in jobrec['qcvars'].items():
+                if not isinstance(qca.data, np.ndarray):
+                    core.set_variable(k, qca.data)
+
+            if self.fctldash in ['hf3c', 'pbeh3c']:
+                gcp_part = gcp.run_gcp(molecule, self.fctldash, verbose=False, dertype=1)
```

Comment:
I suppose while there are only 2 functionals using `gcp`, it's not really a big deal, but it wouldn't hurt to standardize this by adding a `"gcp"` key to the `dict_func` parsers...

Reply:
Yeah, I need to rewrite the gcp interface. Unlike in dftd3, where everything is controlled through `.dftd3_params` and internal-to-dftd3-exe definitions of, say, PBE, are ignored, gcp just [passes the keyword through](https://github.com/psi4/psi4/blob/master/psi4/driver/qcdb/interface_gcp.py#L215). So parameters added to the `dict_func` parsers would have no place to go.

You're completely right, though, about explicit parameter handling (assuming gcp params are as well-defined as dftd3) and a place in the dft dict definition being the desired end.

Reply:
the standard/short version of the gcp parameter file includes the name of basis set, and otherwise 4 numbers. The basis set related values can be supplied in an 'extended' version, which is automatically detected by gcp.

However, hf3c and pbeh3c require additional ingredients which are currently only detected by the `-level` keyword.

Reply:
thanks, gtk. As the 3c's were our target cases, explains why (even in my haste) I didn't reuse the `dftd3_param` machinery in gcp. Glad to rework it should need arise.